Add ApiKey Auth support

cfbarbero opened the pull request:
*Issue #, if available:*
Fixes #867 

*Description of changes:*
Update the `Auth` property to include a new sub-property `ApiKeyRequired: <boolean>`. This will be supported on both the [Api Auth Method](https://github.com/cfbarbero/serverless-application-model/blob/master/versions/2016-10-31.md#api-auth-object) and the [Function Auth Method](https://github.com/cfbarbero/serverless-application-model/blob/master/versions/2016-10-31.md#function-auth-object). The function level specification will override the API level.

For example:
```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: API Gateway with ApiKey Auth
Resources:
  MyApi:
    Type: AWS::Serverless::Api
    Properties:
      StageName: Prod
      Auth:
        ApiKeyRequired: true

  MyFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: .
      Handler: index.handler
      Runtime: nodejs8.10
      Events:
        ApiKeyFalse:
          Type: Api
          Properties:
            RestApiId: !Ref MyApi
            Path: /
            Method: get
            # This is a public endpoint
            Auth:
              ApiKeyRequired: false
        ApiKeyTrue:
          Type: Api
          Properties:
            RestApiId: !Ref MyApi
            Path: /
            Method: get
            Auth:
              ApiKeyRequired: true
        ApiKeyDefault:
          Type: Api
          Properties:
            RestApiId: !Ref MyApi
            Path: /other
            Method: get
```

This is roughly based upon #444.  It implements part of #547. 

In implementing this change, I refactored the support for the Authorizers in the [swagger.py](samtranslator/swagger/swagger.py) file. ~~By doing this, it changed the order in which some of the generated swagger properties are translated and there by required updates to the following translator test output files:~~
- ~~tests/translator/output/api_with_aws_iam_auth_overrides.json~~
- ~~tests/translator/output/api_with_default_aws_iam_auth.json~~

*Description of how you validated changes:*
Added additional translator tests:
- tests/translator/input/api_with_apikey_default_override.yaml
- tests/translator/input/api_with_apikey_required.yaml
Updated existing translator tests:
- tests/translator/input/globals_for_api.yaml
- tests/translator/input/implicit_api_with_auth_and_conditions_max.yaml


*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [x] Update documentation


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

brettstack commented:
>  it changed the order in which some of the generated swagger properties

Please try to maintain the Swagger integrity as SAM will create an APIGW Deployment when it detects Swagger has changed and we strongly try not to do an APIGW Deployment unless the user actually made a change to their API.

cfbarbero commented:
@brettstack 
 I have fixed the bug that caused the AWS_IAM security definition to get reordered.  I have reverted the test output files for the 2 tests (across all 3 regions) that I had changed to accomodate the updated ordering and they are passing.

Let me know if there's anything else you'd like me to address.

praneetap commented on samtranslator/model/eventsources/push.py:
```diff
@@ -562,6 +562,16 @@ def _add_swagger_integration(self, api, function):
 
+            apikey_required_setting = self.Auth.get('ApiKeyRequired')
+            apikey_required_setting_is_false = apikey_required_setting is not None and not apikey_required_setting
+            if apikey_required_setting_is_false and not api_auth.get('ApiKeyRequired'):
```
Throwing an exception here is not required, as this would be a no-op on api gw.

cfbarbero replied:
Like the Authorizers implementation, I am storing the explicit ApiKeyRequired: False for a function in the security block for that function as `api_key_false`. 
```
"security": { "api_key_false": [] }
```
This security option gets cleaned up by the swagger.set_path_default_apikey_required method, however, this is only called when there is a default ApiKeyRequired setting on the Serverless API.  If there is no default setting (or no explicit Serverless API) then we never loop back and cleanup the `api_key_false` security setting.

I assumed we didn't want to spit out this security setting, however, I'll defer to your judgement.  I'm also open to other ideas on this.

praneetap commented on samtranslator/model/api/api_generator.py:
```diff
@@ -274,9 +274,13 @@ def _add_auth(self):
+            self._set_default_apikey_required(swagger_editor)
+
         # Assign the Swagger back to template
         self.definition_body = swagger_editor.swagger
```
Could you please rebase and add/update a test to ensure this works with openapi 3 as well.

cfbarbero replied:
@praneetap , I have rebased.  I am happy to add a test for openapi 3.  Can you provide a little direction on that?  Are there other tests I can refer to as an example?

cfbarbero replied:
@praneetap , I've added a couple of openapi3 translator tests.  Please let me know if there's anything else you'd like to see on this.

New Test:
- api_with_apikey_required_openapi_3

Updated Tests:
- api_with_auth_all_maximum_openapi_3
- api_with_auth_all_maximum

praneetap replied:
Yes, will be happy to! OpenApi tests are run [here](https://github.com/awslabs/serverless-application-model/blob/develop/tests/translator/test_translator.py#L345).
You can find more examples of input templates [here](https://github.com/awslabs/serverless-application-model/tree/develop/tests/translator/input) if you search for "openapi".

cfbarbero commented on samtranslator/model/api/api_generator.py:
```diff
@@ -108,6 +109,8 @@ def _construct_rest_api(self):
         self._add_cors()
         self._add_auth()
+        # Post Process OpenApi Auth Settings
+        self.definition_body = self._openapi_auth_postprocess(self.definition_body)
```
@praneetap In order to support the ApiKeyRequired setting at the Function level (without requiring an explicit setting on the API), I had to move the `_openapi_auth_postprocess` method call up a level to ensure it was called on every template reqardless of if there's an explicit API Auth setting.  If you have another suggestion I'm happy to change.

cfbarbero commented:
@praneetap I have addressed your PR comments, this is ready for review again.

cfbarbero commented on samtranslator/model/api/api_generator.py:
```diff
@@ -115,6 +116,8 @@ def _construct_rest_api(self):
             rest_api.BodyS3Location = self._construct_body_s3_dict()
         elif self.definition_body:
+            # # Post Process OpenApi Auth Settings
+            self.definition_body = self._openapi_auth_postprocess(self.definition_body)
```
@praneetap In order to support the ApiKeyRequired setting at the Function level (without requiring an explicit setting on the API), I had to move the _openapi_auth_postprocess method call up a level to ensure it was called on every template reqardless of if there's an explicit API Auth setting. If you have another suggestion I'm happy to change.

jlhood commented:
@emonhaider We are working on the v1.14.0 release now. We can't give exact dates, but it's coming soon.
